compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/Env.hs

```diff
@@ -58,14 +62,21 @@ getWorld = view world
 
 -- | Type class constraint capturing the needed monadic effects for the
 -- functions manipulating the type checker environment.
-type MonadGamma m = (MonadError Error m, MonadReader Gamma m, MonadState [Warning] m)
+type MonadGamma m = MonadGammaF Gamma m
+type MonadGammaF gamma m = ReaderT gamma (StateT [Warning] (Either Error)) ~ m
 
 runGamma
   :: World
   -> Version
-  -> StateT [Warning] (ReaderT Gamma (Either Error)) a
+  -> ReaderT Gamma (StateT [Warning] (Either Error)) a
```

Comment:
We don't need to, but the implementation of `runGammaUnderUpgrades` and potential other `MonadGammaF` is simpler because `gamma` is what varies, not the warning state.